pkg/cmd/save.go

```diff
@@ -1,43 +1,127 @@
 package cmd
 
 import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
 	"strings"
 
 	"github.com/appvia/artefactor/pkg/docker"
+	"github.com/appvia/artefactor/pkg/git"
+	"github.com/appvia/artefactor/pkg/hashcache"
+	"github.com/appvia/artefactor/pkg/util"
+	"github.com/appvia/artefactor/pkg/web"
+	"github.com/pkg/errors"
 	"github.com/spf13/cobra"
 )
 
 // SaveCommand is the sub command syntax
-const SaveCommand string = "save"
+const (
+	SaveCommand     string = "save"
+	SaveDirMetaFile string = "saveDir.meta"
+)
 
-// cleanupCmd represents the version command
+// saveCmd represents the version command
 var saveCmd = &cobra.Command{
 	Use:   SaveCommand,
 	Short: "saves artefact(s)",
 	Long:  "will save artefact(s) to file(s)",
-	Run: func(c *cobra.Command, args []string) {
-		save(c)
+	RunE: func(c *cobra.Command, args []string) error {
+		return save(c)
 	},
 }
 
 func init() {
-	saveCmd.PersistentFlags().String(
-		"artifacts-dir",
-		"./downloads",
-		"a location to save / load all artefacts")
+	addFlagWithEnvDefault(
+		saveCmd,
+		FlagArchiveDir,
+		DefaultArchiveDir,
+		"a location to save artefacts to and publish from")
 
 	RootCmd.AddCommand(saveCmd)
 }
 
-func save(c *cobra.Command) {
+func save(c *cobra.Command) error {
+	common(c)
+	// Record where the archives should be stored
+	saveDir := c.Flag(FlagArchiveDir).Value.String()
+	if err := saveSavedPath(saveDir); err != nil {
+		return err
+	}
+
 	// First save docker images
-	images := strings.Split(c.Flag(FlagDockerImages).Value.String(), "\n")
+	images := strings.Fields(c.Flag(FlagDockerImages).Value.String())
 	for _, image := range images {
-		docker.Save(
-			image,
-			c.Flag(FlagArchiveDir).Value.String())
+		if err := docker.Save(image, saveDir); err != nil {
+			return err
+		}
+	}
+
+	// Next save any git repos
+	gitRepos := strings.Fields(c.Flag(FlagGitRepos).Value.String())
+	for _, repo := range gitRepos {
+		if err := git.Archive(repo, saveDir); err != nil {
+			return err
```

Comment:
format error with context

Reply:
Indeed, nice to be nice - improved all top level error messages.